### PR TITLE
Finish off container documentation

### PR DESCRIPTION
**Pull request by markstory**
Write the remaining first pass on the container documentation. I've also documented the ServiceProvider interface we will have but hasn't been written yet.

### CODE REVIEW

File: en/development/dependency-injection.rst [hunk 1]
```diff
@@ -136,10 +136,71 @@ configuration.
 Service Providers
 =================
 
-* What is a service provider.
-* Why use service providers
+Service providers allow you to group of related services together helping you
+organize your services. ServiceProviders also help increase your application's
+performance as services defined in ServiceProviders are lazily registered after
```
Comment (othercorey) [inline en/development/dependency-injection.rst:141]:
```suggestion
Service providers allow you to group of related services together helping you
organize your services. Service providers also help increase your application's
performance as services defined in ServiceProvider are lazily registered after
```
Not sure what way is most consistent.

File: en/development/dependency-injection.rst [hunk 2]
```diff
@@ -136,10 +136,71 @@ configuration.
...
+        ];
+
+        public function services($container)
+        {
+            $container->add(StripService::class);
+            $container->add('configKey', 'some value');
+        }
+    }
+
+ServiceProviders use their ``services()`` method to define all the services they
```
Comment (othercorey) [inline en/development/dependency-injection.rst]:
```suggestion
Service providers use their ``services()`` method to define all the services they
```

File: en/development/dependency-injection.rst [hunk 3]
```diff
@@ -136,10 +136,71 @@ configuration.
 Service Providers
 =================
 
-* What is a service provider.
-* Why use service providers
+Service providers allow you to group of related services together helping you
+organize your services. ServiceProviders also help increase your application's
+performance as services defined in ServiceProviders are lazily registered after
+their first use.
```
Comment (othercorey) [inline en/development/dependency-injection.rst]:
```suggestion
organize your services. Service providers can help increase your application's
performance as defined services are lazily registered after
their first use.
```

File: en/development/dependency-injection.rst [hunk 4]
```diff
@@ -136,10 +136,71 @@ configuration.
...
+
+ServiceProviders use their ``services()`` method to define all the services they
+will provide. Additionally those services  **must be** defined in the ``$provides``
+property. Failing to include a service in the ``$provides`` property will result
+in it not be loadable from the container.
+
+Using Service Providers
+-----------------------
+
+To load a ServiceProvider add it into the container using the
```
Comment (othercorey) [inline en/development/dependency-injection.rst]:
```suggestion
To load a service provider add it into the container using the
```

File: en/development/dependency-injection.rst [hunk 5]
```diff
@@ -136,10 +136,71 @@ configuration.
...
+``addServiceProvider()`` method::
+
+    $container->addServiceProvider(new BillingServiceProvider());
+
+Bootable ServiceProviders
+-------------------------
+
+If your service provider needs to run logic when it is added to the container
```
Comment (othercorey) [inline en/development/dependency-injection.rst:184]:
```suggestion
If your service provider needs to run logic when it is added to the container,
```

File: en/development/dependency-injection.rst [hunk 6]
```diff
@@ -136,10 +136,71 @@ configuration.
...
+you can implement the ``bootstrap()`` method. This situation can come up when you
```
Comment (othercorey) [inline en/development/dependency-injection.rst]:
```suggestion
If your service provider needs to run logic when it is added to the container,
you can implement the ``bootstrap()`` method. This situation can come up when your
```

File: en/development/dependency-injection.rst [hunk 7]
```diff
@@ -136,10 +136,71 @@ configuration.
 Service Providers
 =================
 
-* What is a service provider.
-* Why use service providers
+Service providers allow you to group of related services together helping you
```
Comment (jadb) [inline en/development/dependency-injection.rst]:
```suggestion
Service providers allow you to group related services together helping you
```

**Pull request merged**